crates/turborepo-lib/src/run/mod.rs

```diff
@@ -566,4 +560,50 @@ impl<'a> Run<'a> {
 
         Ok((global_hash, task_hash_tracker))
     }
+
+    fn build_engine(
+        &self,
+        pkg_dep_graph: &PackageGraph,
+        opts: &Opts,
+        root_turbo_json: &TurboJson,
+        filtered_pkgs: &HashSet<WorkspaceName>,
+    ) -> Result<Engine> {
+        let engine = EngineBuilder::new(
+            &self.base.repo_root,
+            pkg_dep_graph,
+            opts.run_opts.single_package,
+        )
+        .with_root_tasks(root_turbo_json.pipeline.keys().cloned())
+        .with_turbo_jsons(Some(
+            Some((WorkspaceName::Root, root_turbo_json.clone()))
+                .into_iter()
+                .collect(),
+        ))
+        .with_tasks_only(opts.run_opts.only)
+        .with_workspaces(filtered_pkgs.clone().into_iter().collect())
+        .with_tasks(
+            opts.run_opts
+                .tasks
+                .iter()
+                .map(|task| TaskName::from(task.as_str()).into_owned()),
+        )
+        .build()?;
+
+        if !opts.run_opts.parallel {
+            engine
+                .validate(pkg_dep_graph, opts.run_opts.concurrency)
```

Comment:
Just a note that we should fix #4291 with this in the future. I think it's fine to only fix in Rust if we're close